storage/innobase/handler/ha_innodb.cc

```diff
@@ -19989,20 +19999,24 @@ i_s_innodb_sys_virtual,
 i_s_innodb_tablespaces_encryption
 maria_declare_plugin_end;
 
-/** @brief Adjust some InnoDB startup parameters based on file contents
-or innodb_page_size. */
-static
-void
-innodb_params_adjust()
+/** Adjust some InnoDB startup parameters based on the data directory */
+static void innodb_params_adjust()
 {
-	MYSQL_SYSVAR_NAME(max_undo_log_size).max_val
-		= 1ULL << (32U + srv_page_size_shift);
-	MYSQL_SYSVAR_NAME(max_undo_log_size).min_val
-		= MYSQL_SYSVAR_NAME(max_undo_log_size).def_val
-		= ulonglong(SRV_UNDO_TABLESPACE_SIZE_IN_PAGES)
-		<< srv_page_size_shift;
-	MYSQL_SYSVAR_NAME(max_undo_log_size).max_val
-		= 1ULL << (32U + srv_page_size_shift);
+  MYSQL_SYSVAR_NAME(max_undo_log_size).max_val=
+    1ULL << (32U + srv_page_size_shift);
+  MYSQL_SYSVAR_NAME(max_undo_log_size).min_val=
+    MYSQL_SYSVAR_NAME(max_undo_log_size).def_val=
+    ulonglong{SRV_UNDO_TABLESPACE_SIZE_IN_PAGES} << srv_page_size_shift;
+  MYSQL_SYSVAR_NAME(max_undo_log_size).max_val=
+    1ULL << (32U + srv_page_size_shift);
+#if defined __linux__ || defined _WIN32
+  uint &min_val= MYSQL_SYSVAR_NAME(log_write_ahead_size).min_val;
+  if (min_val < log_sys.write_size)
+    min_val= log_sys.write_size;
```

Comment:
1. Yes, we probably should change the default value as well, to be consistent. We do not allow `SET GLOBAL innodb_log_write_ahead_size=DEFAULT` yet, but we might later in #3327.
2. Yes, but in `mysql-test/suite/sys_vars/t/sysvars_innodb.test` we filter out this variable because the minimum value depends on the environment. We want the default to be 512 in order not to introduce a performance regression in environments where the underlying block size is 512 bytes. But we must adjust the minimum to not be less than the physical block size.

To my surprise, with this patch:
```diff
diff --git a/mysql-test/suite/sys_vars/t/sysvars_innodb.test b/mysql-test/suite/sys_vars/t/sysvars_innodb.test
index 9e1b1d1a26d..2680e442da4 100644
--- a/mysql-test/suite/sys_vars/t/sysvars_innodb.test
+++ b/mysql-test/suite/sys_vars/t/sysvars_innodb.test
@@ -12,6 +12,5 @@ select VARIABLE_NAME, SESSION_VALUE, DEFAULT_VALUE, VARIABLE_SCOPE, VARIABLE_TYP
     'innodb_evict_tables_on_commit_debug', # one may want to override this
     'innodb_use_native_aio',            # default value depends on OS
     'innodb_log_file_buffering',        # only available on Linux and Windows
-    'innodb_log_write_ahead_size',      # the minimum value is system dependent
     'innodb_buffer_pool_load_pages_abort')            # debug build only, and is only for testing
   order by variable_name;
diff --git a/storage/innobase/os/os0file.cc b/storage/innobase/os/os0file.cc
index 49b243fc96d..c3328dd8963 100644
--- a/storage/innobase/os/os0file.cc
+++ b/storage/innobase/os/os0file.cc
@@ -1108,7 +1108,7 @@ static ATTRIBUTE_COLD bool os_file_log_maybe_unbuffered(const struct stat &st)
       static_cast<int>(sizeof b))
   {
   fallback:
-    log_sys.set_block_size(512);
+    log_sys.set_block_size(2048);
     return false;
   }
   int f= open(b, O_RDONLY);
```
and running with `mysql-test/var` pointing to `/run/user/$UID/var`, I am seeing the following output:
```diff
CURRENT_TEST: sys_vars.sysvars_innodb
--- /mariadb/10.11/mysql-test/suite/sys_vars/r/sysvars_innodb.result	2024-06-27 13:01:40.724565329 +0300
+++ /mariadb/10.11/mysql-test/suite/sys_vars/r/sysvars_innodb.reject	2024-06-27 13:09:20.862229395 +0300
@@ -5,7 +5,6 @@
 'innodb_evict_tables_on_commit_debug', # one may want to override this
 'innodb_use_native_aio',            # default value depends on OS
 'innodb_log_file_buffering',        # only available on Linux and Windows
-'innodb_log_write_ahead_size',      # the minimum value is system dependent
 'innodb_buffer_pool_load_pages_abort')            # debug build only, and is only for testing
 order by variable_name;
 VARIABLE_NAME	INNODB_ADAPTIVE_FLUSHING
@@ -1040,6 +1039,18 @@
 ENUM_VALUE_LIST	NULL
 READ_ONLY	NO
 COMMAND_LINE_ARGUMENT	OPTIONAL
+VARIABLE_NAME	INNODB_LOG_WRITE_AHEAD_SIZE
+SESSION_VALUE	NULL
+DEFAULT_VALUE	512
+VARIABLE_SCOPE	GLOBAL
+VARIABLE_TYPE	INT UNSIGNED
+VARIABLE_COMMENT	Redo log write size to avoid read-on-write; must be a power of two
+NUMERIC_MIN_VALUE	512
+NUMERIC_MAX_VALUE	4096
+NUMERIC_BLOCK_SIZE	1
+ENUM_VALUE_LIST	NULL
+READ_ONLY	YES
+COMMAND_LINE_ARGUMENT	REQUIRED
 VARIABLE_NAME	INNODB_LRU_FLUSH_SIZE
 SESSION_VALUE	NULL
 DEFAULT_VALUE	32
```
I checked with `./mtr --rr` and some `awatch` that the correct value of the `min_val=2048` is being accessed, but we still see the 512 above. Same for the current value (but I think that the above test is filtering it out from the result):
```
Thread 7 hit Hardware access (read/write) watchpoint 1: log_sys.write_size

Value = 2048
0x0000557e2279b9ec in get_one_variable (thd=thd@entry=0x7f556c000f98, 
    variable=variable@entry=0x7f556c129180, 
    value_type=value_type@entry=SHOW_OPT_GLOBAL, show_type=SHOW_UINT, 
    show_type@entry=SHOW_SYS, status_var=status_var@entry=0x0, 
    charset=charset@entry=0x7f5597a7a6d0, buff=0x7f5597a7a740 "262144", 
    length=0x7f5597a7a6d8) at /mariadb/10.11/sql/sql_show.cc:3728
3728	    end= int10_to_str((long) *value.as_uint, buff, 10);
```
The `def_val` would only be accessed in `plugin_opt_set_limits()` and not thereafter. I guess that there is another copy of this data that `information_schema.system_variables` is reflecting.

Reply:
I set a breakpoint to `fill_sysvars` with the condition `$_streq(var->name.str,"innodb_log_write_ahead_size")`. I subsequently set a watchpoint on `var->option.min_val` to find out that the data had indeed been copied in `plugin_opt_set_limits()` before InnoDB was even started:
```c++
void plugin_opt_set_limits(struct my_option *options,
                           const struct st_mysql_sys_var *opt)
{
  options->sub_size= 0;

  switch (opt->flags & (PLUGIN_VAR_TYPEMASK |
                        PLUGIN_VAR_UNSIGNED | PLUGIN_VAR_THDLOCAL)) {
  /* global system variables */
  case PLUGIN_VAR_INT:
    OPTION_SET_LIMITS(GET_INT, options, (sysvar_int_t*) opt);
    break;
  case PLUGIN_VAR_INT | PLUGIN_VAR_UNSIGNED:
    OPTION_SET_LIMITS(GET_UINT, options, (sysvar_uint_t*) opt);
    break;
```
So, it looks like this adjustment in `innodb_params_adjust()` is useless as it is for now. I will add a source code comment or an `#if 0` block and stop filtering out `innodb_log_write_ahead_size` in `mysql-test/suite/sys_vars/t/sysvars_innodb.test`.